### PR TITLE
Fix static linkage cases and NO_DISTRIBUTED=1 + CUDA (#16705)

### LINKED ISSUE #14805: libcudart.so not found when compiling with NO_DISTRIBUTED=1
When compiling PyTorch from sources (a80a46a6d01b792399d5860e00923d096399952f) with the flag NO_DISTRIBUTED=1 it ends up with the Following error : 

`couldn't find libcudart. Make sure CUDA libraries are installed in adefault location, or that they're in LD_LIBRARY_PATH` comming from there : https://github.com/pytorch/pytorch/blob/master/torch/cuda/__init__.py#L66

Compiling with Distributed support makes the error disappear.

Happened on Fedora 29, using GCC/G++ 7.3.1 with CUDA 9.2 / CuDNN 7.4.1 / GTX 1080 Ti

### PR DESCRIPTION
**Pull request by soumith**
Attempt #2 (attempt 1 is https://github.com/pytorch/pytorch/pull/16705 and got reverted because of CI failures)

Fixes https://github.com/pytorch/pytorch/issues/14805

### CODE REVIEW

**Comment by mfuntowicz** [comment]
I checked out your branch and it seems to be ok (at least for me).

Thanks !

File: c10/cuda/CUDAException.h [hunk 1]
```diff
@@ -14,6 +14,7 @@
   do {                                                     \
     cudaError_t __err = EXPR;                              \
     if (__err != cudaSuccess) {                            \
+      cudaGetLastError();				   \
```
Comment (soumith) [inline c10/cuda/CUDAException.h]: @ngimel can you review that this is okay?

Without this, a previously raised error was still lingering and falsely being triggered for a subsequent CUDA call. @colesbury suggested that this is the right thing to do.

Comment (ngimel) [reply c10/cuda/CUDAException.h]: Yep, it is. It won't work if the error is sticky and context was corrupted, but for other cases if you can recover from RuntimeError thrown by AT_ERROR, you are good to go.

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**